# Case 1 Heart Disease Classification, model nro.2
Wille Tuovinen and Eemil Rantanen, 
4.2.2019, 
Helsinki Metropolia University of Applied Sciences
 
# Reason why we are building this Neural Network
Objective of this document was to lear usage of librarys and building a Neural Network. Goal was allso to lear and test bifferent model architecture's and visualize outputs with matplotlib. 
Here we tried another way to create NN, but our end-output isn't quite accurate/clear to us.

___________________________________________________________________________________________________________________________
___________________________________________________________________________________________________________________________

### Importing required librarys and data. 
Pandas DataFrame gives massive functionality to work on data thus, here we are using pandas to import data.

In [2]:
# calling required libraries
import numpy as np
import time
import warnings
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.utils import shuffle
from sklearn.preprocessing import normalize
from sklearn.metrics import confusion_matrix, precision_recall_fscore_support
from keras.utils import to_categorical
from keras import models, layers
from keras.utils import to_categorical
from keras.models import Sequential
from keras.layers import Dense
from sklearn.metrics import confusion_matrix
clasifier = Sequential()

Using TensorFlow backend.


In [5]:
data = 'https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data'
df = pd.read_csv(data, sep=',' , header = None)

# Let's convert some missplaysed "?" strings to lets say, to 0, with this awfull "monster" of a code.
df = df.fillna(df.median())
df = df.fillna(0.0)
df = df.replace("?", 0.0)

# naming columns
df.columns = ["age", "sex", "cp", "treshbps", "chol", "fbs", "restecg", "thalach",
              "exang", "oldpeak", "slope", "ca", "thal", "num"]

# here we change values higher that 0.0 to 2.0, and values equal to  0.0, to 1.0
df.loc[df.num > 0.0, 'num'] = 2.0
df.loc[df.num <= 0.0, 'num'] = 1.0

# rename axis
df.index.name = 'id'   

df['num'] = pd.to_numeric(df['num']).astype(float)

# Randomize rows
df = shuffle(df)
df.head()
# Here you see that the data is randomized

age  sex   cp  treshbps   chol  fbs  restecg  thalach  exang  oldpeak  \
id                                                                            
215  56.0  1.0  1.0     120.0  193.0  0.0      2.0    162.0    0.0      1.9   
206  58.0  1.0  4.0     128.0  259.0  0.0      2.0    130.0    1.0      3.0   
46   51.0  1.0  3.0     110.0  175.0  0.0      0.0    123.0    0.0      0.6   
162  54.0  0.0  3.0     110.0  214.0  0.0      0.0    158.0    0.0      1.6   
186  42.0  1.0  3.0     120.0  240.0  1.0      0.0    194.0    0.0      0.8   

     slope   ca thal  num  
id                         
215    2.0  0.0  7.0  1.0  
206    2.0  2.0  7.0  2.0  
46     1.0  0.0  3.0  1.0  
162    2.0  0.0  3.0  1.0  
186    3.0  0.0  7.0  1.0

### Here we generate .csv from the data

In [77]:
df.to_csv('Case1.2_.csv')

#If you look at the CSV there, you should see it has the headers. 
#df.to_csv('Case1.2_wille.csv', header=False)

# Or you whatn to save file as HTML
#df.to_html('Case1.2_wille.html')


### Create matrix of features and matrix of target variable

In [7]:
# Create matrix of features and matrix of target variable
x = df.iloc[:, 0:13].values
y = df.iloc[:, 13].values

### We will make use of ScikitLearn’s ‘train_test_split’ function to divide our data 

In [79]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2)

In [80]:
# Feature Scaling
#from sklearn.preprocessing import StandardScaler
#sc = StandardScaler()
#x_train = sc.fit_transform(x_train)
#x_tst = sc.transform(x_test)
x_train

array([[37.0, 0.0, 3.0, ..., 1.0, '0.0', '3.0'],
       [57.0, 1.0, 4.0, ..., 2.0, '3.0', '7.0'],
       [58.0, 0.0, 1.0, ..., 1.0, '0.0', '3.0'],
       ...,
       [64.0, 1.0, 1.0, ..., 2.0, '0.0', '3.0'],
       [52.0, 1.0, 3.0, ..., 1.0, 0.0, '3.0'],
       [57.0, 1.0, 4.0, ..., 2.0, '0.0', '6.0']], dtype=object)

## Starting to built NN
### Initializing Neural Network

Our first parameter is output_dim. It is simply the number of nodes you want to add to this layer. 
init is the initialization of Stochastic Gradient Decent. In Neural Network we need to assign weights to each mode which is nothing but importance of that node. 

At the time of initialization, weights should be close to 0 and we will randomly initialize weights using uniform function. input_dim parameter is needed only for first layer as model doesn’t know the number of our input variables. 
As we knove the total number of input variables are 13. In the second layer model automatically knows the number of input variable from the first hidden layer.

In upcoming code we recognized that syntax has changed during last few years. for example "init" is nowdays replaced with "kernel_initializer". There wasnt big problems with old syntax. while deployng code, it gave warnings, that the code is updated and replaced with....blaa blaa blaa,

In [81]:
# Adding the input layer and the first hidden layer
#clasifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu', input_dim = 13))            THIS IS GIVING ERRORS
clasifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 13))

# Adding the second hidden layer
#clasifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu'))                             THIS IS GIVING ERRORS
clasifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))

# Adding the output layer
#clasifier.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))                           THESE ARE GIVING ERRORS
clasifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid' ))


### Compiling Neural Network

First argument is Optimizer, this is nothing but the algorithm you wanna use to find optimal set of weights

In [82]:
# Now here's the Deep Learning Neural Network model.
clasifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
clasifier.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_20 (Dense)             (None, 6)                 84        
_________________________________________________________________
dense_21 (Dense)             (None, 6)                 42        
_________________________________________________________________
dense_22 (Dense)             (None, 1)                 7         
Total params: 133
Trainable params: 133
Non-trainable params: 0
_________________________________________________________________


### Fitting our model
Here we train our model"classifier(typoed"clasifier :D")" on training data but still one thing is remaining.
Here we use fit method to the fit our model.
Batch size is used to specify the number of observation after which you want to update weight. 
Epoch is nothing but the total number of iterations.

In [83]:
# Fitting our model
trainStart = time.time()
clasifier.fit(x_train, y_train, epochs=100, batch_size=10)

trainEnd = time.time()
print(" ")
print('Elapsed time: {:.2f} seconds'.format(trainEnd - trainStart))

Epoch 1/100
242/242 [==============================] - 1s 2ms/step - loss: 0.6838 - acc: 0.4215 
Epoch 2/100
242/242 [==============================] - 0s 157us/step - loss: 0.6583 - acc: 0.5331
Epoch 3/100
242/242 [==============================] - 0s 152us/step - loss: 0.6349 - acc: 0.5331
Epoch 4/100
242/242 [==============================] - 0s 223us/step - loss: 0.6122 - acc: 0.5331
Epoch 5/100
242/242 [==============================] - 0s 136us/step - loss: 0.5893 - acc: 0.5331
Epoch 6/100
242/242 [==============================] - 0s 190us/step - loss: 0.5663 - acc: 0.5331
Epoch 7/100
242/242 [==============================] - 0s 165us/step - loss: 0.5439 - acc: 0.5331
Epoch 8/100
242/242 [==============================] - 0s 157us/step - loss: 0.5217 - acc: 0.5331
Epoch 9/100
242/242 [==============================] - 0s 169us/step - loss: 0.4995 - acc: 0.5331
Epoch 10/100
242/242 [==============================] - 0s 157us/step - loss: 0.4773 - acc: 0.5331
Epoch 11/100
242/242

Epoch 83/100
242/242 [==============================] - 0s 152us/step - loss: -0.7079 - acc: 0.5331
Epoch 84/100
242/242 [==============================] - 0s 210us/step - loss: -0.7206 - acc: 0.5331
Epoch 85/100
242/242 [==============================] - 0s 157us/step - loss: -0.7330 - acc: 0.5331
Epoch 86/100
242/242 [==============================] - 0s 132us/step - loss: -0.7455 - acc: 0.5331
Epoch 87/100
242/242 [==============================] - 0s 148us/step - loss: -0.7577 - acc: 0.5331
Epoch 88/100
242/242 [==============================] - 0s 132us/step - loss: -0.7697 - acc: 0.5331
Epoch 89/100
242/242 [==============================] - 0s 144us/step - loss: -0.7821 - acc: 0.5331
Epoch 90/100
242/242 [==============================] - 0s 144us/step - loss: -0.7942 - acc: 0.5331
Epoch 91/100
242/242 [==============================] - 0s 144us/step - loss: -0.8063 - acc: 0.5331
Epoch 92/100
242/242 [==============================] - 0s 152us/step - loss: -0.8186 - acc: 0.5331


### Predicting the test set result. 
If we print "y_pred" , we would get 1 column fild with whole much of "trues".
We didint quit understand why accuracy is same in every epoch. Though "loss" changes in epoch's.

In [86]:
# Predicting the Test set results
y_pred = clasifier.predict(x_test)
y_pred = (y_pred > 0.5)

# This is the final step where we are evaluating our model performance. 
We have original results and thus we can build confusion matrix to check the accuracy of model.

In [85]:
# Creating the Confusion Matrix

#cm = confusion_matrix(y_test, y_pred)
a = confusion_matrix(y_test, y_pred)
df.to_csv('Case1.2_confusion_matrix.csv')
a


array([[35,  0],
       [26,  0]], dtype=int64)

#### So the Accuracy of our model can be calculated as:

Accuracy= 35x + 0y / 61z = 0.5737704918032787

We got 57% accuracy.